src/webgpu/api/operation/command_buffer/queries/occlusionQuery.spec.ts

```diff
@@ -0,0 +1,1076 @@
+      },
+      (passed, queryIndex) => {
+        const drawMask = queryIndex % 2 ? 2 : 4;
+        const expectPassed = !!(sampleMask & drawMask);
+        t.expect(
+          !!passed === expectPassed,
+          `queryIndex: ${queryIndex}, was: ${!!passed}, expected: ${expectPassed}, ${name}`
+        );
+      }
+    );
+  });
+
+g.test('occlusion_query,alpha_to_coverage')
+  .desc(
+    `
+      Test beginOcclusionQuery/endOcclusionQuery using alphaToCoverage to occlude
+
+      Set alpha to 0.25, draw quads in 4 corners of texel. Some should be culled.
+    `
+  )
+  .params(kQueryTestBaseParams.combine('alpha', [0, 0.25, 0.5, 0.75, 1.0]))
+  .fn(async t => {
+    const { writeMask, renderMode, bufferOffset, querySetOffset, alpha } = t.params;
+    const kNumQueries = 32;
+    const sampleCount = 4;
+    const resources = t.setup({
+      writeMask,
+      renderMode,
+      bufferOffset,
+      querySetOffset,
+      numQueries: kNumQueries,
+      sampleCount,
+      alpha,
+    });
+    const { occlusionQuerySet, renderTargetTexture, pipeline } = resources;
+
+    const createQuad = (xOffset: number, yOffset: number) => {
+      // prettier-ignore
+      return t.createVertexBuffer(new Float32Array([
+        xOffset + 0   , yOffset + 0   , 0,
+        xOffset + 0.25, yOffset + 0   , 0,
+        xOffset + 0   , yOffset + 0.25, 0,
+        xOffset + 0   , yOffset + 0.25, 0,
+        xOffset + 0.25, yOffset + 0   , 0,
+        xOffset + 0.25, yOffset + 0.25, 0,
+      ]));
+    };
+
+    const vertexBuffers = [
+      createQuad(0, 0),
+      createQuad(0.25, 0),
+      createQuad(0, 0.25),
+      createQuad(0.25, 0.25),
+    ];
+
+    const multisampleRenderTarget = t.createTexture({
+      size: kTextureSize,
+      format: 'rgba8unorm',
+      usage: GPUTextureUsage.RENDER_ATTACHMENT,
+      sampleCount,
+    });
+
+    const renderPassDescriptor: GPURenderPassDescriptor = {
+      colorAttachments: [
+        {
+          view: multisampleRenderTarget.createView(),
+          resolveTarget: renderTargetTexture.createView(),
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+      occlusionQuerySet,
+    };
+
+    const numPassedPerGroup = new Array(kNumQueries / 4).fill(0);
+    await t.runQueryTest(
+      resources,
+      renderPassDescriptor,
+      (helper, queryIndex) => {
+        const queryHelper = helper.beginOcclusionQuery(queryIndex);
+        queryHelper.setPipeline(pipeline);
+        queryHelper.setVertexBuffer(vertexBuffers[queryIndex % 4]);
+        queryHelper.draw(6);
+        queryHelper.end();
+      },
+      (passed, queryIndex) => {
+        numPassedPerGroup[(queryIndex / 4) | 0] += passed ? 1 : 0;
+      }
+    );
+
+    const expected = (alpha / 0.25) | 0;
+    numPassedPerGroup.forEach((numPassed, queryGroup) => {
+      t.expect(
+        numPassed === expected,
+        `queryGroup: ${queryGroup}, was: ${numPassed}, expected: ${expected}`
+      );
+    });
+  });
```

Comment:
Wow, is this condition going to be too tight? I'm wondering if its possible for like small deviations from the sampling to cause like maybe a +-1 or something here? (I'm not too familiar with how strict we enforce these both on the API level and how it works in hardware...)

Reply:
Yes! After reading a bit more about what alpha to coverage is I think you are right that this should be exact. Thanks!